contracts/ATM/ATMGovernance.sol

```diff
@@ -0,0 +1,281 @@
+    ) external onlySigner() {
+        require(newProvider.isContract(), "DATA_PROVIDER_MUST_BE_A_CONTRACT");
+        address oldProvider = dataProviders[dataTypeIndex][oldProviderIndex];
+        require(oldProvider != newProvider, "DATA_PROVIDER_SAME_OLD");
+        dataProviders[dataTypeIndex][oldProviderIndex] = newProvider;
+        emit DataProviderUpdated(msg.sender, dataTypeIndex, oldProvider, newProvider);
+    }
+
+    /**
+        @notice Removes an existing Data Provider on a specific Data Type array.
+        @param dataTypeIndex array index for this Data Type.
+        @param dataProviderIndex data provider index.
+     */
+    function removeDataProvider(uint8 dataTypeIndex, uint256 dataProviderIndex)
+        external
+        onlySigner()
+    {
+        require(
+            dataProviders[dataTypeIndex].length > dataProviderIndex,
+            "DATA_PROVIDER_OUT_RANGE"
+        );
+        address oldDataProvider = dataProviders[dataTypeIndex][dataProviderIndex];
+        dataProviders[dataTypeIndex].removeAt(dataProviderIndex);
+        emit DataProviderRemoved(
+            msg.sender,
+            dataTypeIndex,
+            dataProviderIndex,
+            oldDataProvider
+        );
+    }
+
+    /**
+        @notice Sets the CRA - Credit Risk Algorithm to be used on this specific ATM.
+                CRA is represented by a Github commit hash of the newly proposed algorithm.
+     */
+    function setCRA(string calldata _cra) external onlySigner() {
+        bytes memory tempEmptyStringTest = bytes(_cra);
+        require(tempEmptyStringTest.length > 0, "CRA_CANT_BE_EMPTY");
+        require(
+            keccak256(abi.encodePacked(cra)) != keccak256(abi.encodePacked(_cra)),
+            "CRA_SAME_AS_OLD"
+        );
+        cra = _cra;
+        emit CRASet(msg.sender, cra);
+    }
+
+    /* External Constant functions */
+
+    /**
+        @notice Returns a General Setting value from this ATM.
+        @param settingName name of the setting to be returned.
+     */
+    function getGeneralSetting(bytes32 settingName) external view returns (uint256) {
+        return generalSettings[settingName];
+    }
+
+    /**
+        @notice Returns an existing Asset Setting value from a specific Market on this ATM.
+        @param asset market specific asset address.
+        @param settingName name of the setting to be returned.
+     */
+    function getAssetMarketSetting(address asset, bytes32 settingName)
+        external
+        view
+        returns (uint256)
+    {
+        return assetMarketSettings[asset][settingName];
+    }
+
+    /**
+        @notice Returns a Data Provider on a specific Data Type array.
+        @param dataTypeIndex array index for this Data Type.
+        @param dataProviderIndex data provider index number.
+     */
+    function getDataProvider(uint8 dataTypeIndex, uint256 dataProviderIndex)
+        external
+        view
+        returns (address)
+    {
+        require(
```

Comment:
I think we don't need to validate it here (view function). In case the index is out of range it should return 0x0.